searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_job_take2.cpp

```diff
@@ -46,14 +47,19 @@ CompactionJob::moveDocument(const search::DocumentMetaData & meta, std::shared_p
     // The real lid must be sampled in the master thread.
     //TODO remove target lid from createMoveOperation interface
     auto op = _handler.createMoveOperation(meta, 0);
-    if (!op) return;
+    if (!op || !op->getDocument()) return;
+    // Early detection and force md5 calculation outside of master thread
+    if (meta.gid != op->getDocument()->getId().getGlobalId()) return;
 
     _master.execute(makeLambdaTask([this, metaThen=meta, moveOp=std::move(op), onDone=std::move(context)]() {
         search::DocumentMetaData metaNow = _handler.getMetaData(metaThen.lid);
         if (metaNow.lid != metaThen.lid) return;
         if (metaNow.bucketId != metaThen.bucketId) return;
+        if (metaNow.gid != moveOp->getDocument()->getId().getGlobalId()) return;
 
-        moveOp->setTargetLid(_handler.getLidStatus().getLowestFreeLid());
+        uint32_t lowestLid = _handler.getLidStatus().getLowestFreeLid();
+        if (lowestLid >= metaNow.lid) return;
```

Comment:
Just a general comment: We should add test cases for all the return statements in this lambda.